typescript/sdk/src/token/read.ts

```diff
@@ -0,0 +1,138 @@
+import { ethers, providers } from 'ethers';
+
+import {
+  ERC20__factory,
+  HypERC20Collateral__factory,
+} from '@hyperlane-xyz/core';
+import { ERC20Metadata, ERC20RouterConfig } from '@hyperlane-xyz/sdk';
+import { Address } from '@hyperlane-xyz/utils';
+
+import { DEFAULT_CONTRACT_READ_CONCURRENCY } from '../consts/crud.js';
+import { EvmHookReader } from '../hook/read.js';
+import { EvmIsmReader } from '../ism/read.js';
+import { MultiProvider } from '../providers/MultiProvider.js';
+import { ChainName } from '../types.js';
+
+type WarpRouteBaseMetadata = Record<
+  | 'mailbox'
+  | 'owner'
+  | 'wrappedTokenAddress'
+  | 'hook'
+  | 'interchainSecurityModule',
+  string
+>;
+
+interface WarpRouteReader {
+  deriveWarpRouteConfig(address: Address): Promise<Partial<ERC20RouterConfig>>;
+}
```

Comment:
the return type here needs to be more generic as an EvmERC721WarpRouteReader would need to implement the same interface but that returns `ERC721RouterConfig`. i understand this is tricky though for warp vs core/hook/ism because there's so many different variations of a warpconfig

so I suggest we remove this interface and in the warp crud module implementations we manually include `EvmERC20WarpRouterReader` and `EvmERC721WarpRouterReader` instead of a top level interface, as many of these types/routes end up being EVM specific

Reply:
Good point. Will remove interface